### PR TITLE
Several changes in "wp-source" API

### LINKED ISSUE #65: Move pathOrObj and parsePath to wp-source libraries
We need to provide a simple way to move from href to path & page and viceversa. The code should be in `wp-source` because source is responsible for the final urls.

### LINKED ISSUE #102: Instantiate classes on each SSR
In order to ensure that `API` and `Resolver` have a new instance on each SSR we can use a function instead of an object:
```javascript
const WpSource = () => ({
  state...
  actions...
  libraries: {
    api: new Api(),
    resolver: new Resolver(),
    ...
   }
})
```

If `Resolver` or any other library doesn't need a new instance, we can instantiate it in the file:

```js
const resolver: new Resolver();

const WpSource = () => ({
  state...
  actions...
  libraries: {
    api: new Api(),
    resolver,
    ...
   }
})
```

### PR DESCRIPTION
**Pull request by DAreRodz**
Closes #65, #79, #82.
Fixes #102.

### CODE REVIEW

**Comment by luisherranz** [comment]
👏👏👏

File: packages/wp-source/src/actions.ts [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+  fetch: ({ state, libraries }) => async routeOrParams => {
     const { source } = state;
     const { resolver } = libraries.source;
 
     let path: string;
     let page: number;
 
-    if (typeof pathOrObj === "string") {
-      ({ path, page } = parsePath(pathOrObj));
+    if (typeof routeOrParams === "string") {
```
Comment (luisherranz) [inline packages/wp-source/src/actions.ts]: I would get rid of all this and I'd use only something like:
```js
const { path, page } = getParams(routeOrParams);
```

Comment (luisherranz) [reply packages/wp-source/src/actions.ts]: Same here in router:
https://github.com/frontity/frontity/blob/dev/packages/tiny-router/src/actions.ts#L12-L17

```js 
const { path, page } = libraries.source.getParams(routeOrParams);
```

File: packages/wp-source/src/index.ts [hunk 2]
```diff
@@ -15,9 +15,10 @@ const wpSource: WpSource = {
     source: {
       api: new Api(),
       resolver: new Resolver(),
-      populate
+      populate,
+      routeUtils
```
Comment (luisherranz) [inline packages/wp-source/src/index.ts]: For easy access I'd spread `...routeUtils` to get:
`libraries.source.routeToParams` vs `libraries.source.routeUtils.routeToParams`.

File: packages/wp-source/src/state.ts [hunk 3]
```diff
@@ -1,9 +1,15 @@
 import WpSource from "..";
-import { parsePath } from "./utils/parse-path";
+import { normalize, paramsToRoute } from "./libraries/routeUtils";
 
 const state: WpSource["state"]["source"] = {
-  get: ({ state }) => pathOrLink =>
-    state.source.data[parsePath(pathOrLink).path] || {},
+  get: ({ state }) => routeOrObj => {
```
Comment (luisherranz) [inline packages/wp-source/src/state.ts]: I really like to have a `normalize` util but here I'd try to get rid of the ifs delegating that to the utils, doing something like:
```js
const route = getRoute(routeOrParams);
return state.source.data[route] || {};
```

**Review by luisherranz — CHANGES_REQUESTED** [review]
Wow!! Amazing work @DAreRodz!! 👏👏

I just have this comment about the public API of routeUtils:
We have a lot of functions and even the React `<Link>` components with a `routeOrParams` argument, so instead of `routeToParams` and `paramsToRoute` what do you think about a `getParams` and `getRoute` that accept a `routeOrParams`? Or maybe maintain `routeToParams` and `paramsToRoute` but add the other two.

And I'll take a look at the error of `deepMerge` myself.

**Comment by luisherranz** [comment]
@DAreRodz tell me the steps to reproduce the error you saw with `deepmerge` please.

**Comment by DAreRodz** [comment]
In this branch:

- go to `/examples/mars-theme-example`
- run `npx frontity dev`
- access to `http://localhost:3000`

The error will appear in the logs:

```
TypeError: Cannot convert undefined or null to object
      at Function.keys (<anonymous>)
      at mergeObject (webpack-internal:///../../node_modules/deepmerge/dist/umd.js:59:12)
      at deepmerge (webpack-internal:///../../node_modules/deepmerge/dist/umd.js:82:14)
      at state.frontity.packages.forEach.name (webpack-internal:///../../packages/core/src/utils/merge-packages.ts:24:62)
      at Array.forEach (<anonymous>)
      at __webpack_exports__.default (webpack-internal:///../../packages/core/src/utils/merge-packages.ts:20:27)
      at __webpack_exports__.default (webpack-internal:///../../packages/core/src/server/store.ts:17:87)
      at app.use (webpack-internal:///../../packages/core/src/server/index.tsx:86:74)
      at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by DAreRodz** [comment]
Never mind that, I forgot to do `npm i`. Just a missing dependency. 🙄

**Comment by luisherranz** [comment]
No prob!

**Comment by luisherranz** [comment]
I've fixed the bug of `<List>` and it's now included in this PR.

**Comment by DAreRodz** [comment]
@luisherranz, @orballo,  ready for review again!

**Review by luisherranz — APPROVED** [review]
Incredible work @DAreRodz!

We're almost ready for the release candidate 🎉 🎉

**Pull request merged**